src/events.rs

```diff
@@ -0,0 +1,319 @@
+                .map_err(|_e| Error::AddFailure)?;
+
+            self.watched_files.insert(path.clone());
+        }
+
+        Ok(path)
+    }
+
+    fn handle_event(&mut self, event: &mut notify::Event) {
+        let mut retain_vec = vec![];
+
+        for path in &event.paths {
+            if self.pending_watched_files.contains(path) {
+                // TODO: should only do on events that imply file exists
+                let parent = path.parent().expect("Pending watched file needs a parent");
+                self.remove_directory(parent).unwrap();
+                self.pending_watched_files.remove(path);
+                let _res = self.add_file(path);
+            }
+
+            retain_vec.push(self.watched_files.contains(path));
+        }
+
+        let mut i = 0;
+        let paths = &mut event.paths;
+        paths.retain(|_| {
+            let retained = retain_vec[i];
+            i += 1;
+            retained
+        });
+    }
+
+    async fn next_event(&mut self) -> Option<Result<notify::Event, Error>> {
+        self.event_stream.next().await.map(|res| {
+            res.map_err(|e| {
+                match e.kind {
+                    notify::ErrorKind::Io(io_err) => io_err.into(),
+                    _ => {
+                        // Runtime event errors should only be std::io, but
+                        // need to handle this case anyway.
+                        io::Error::new(io::ErrorKind::Other, format!("Event error: {:?}", e)).into()
+                    }
+                }
+            })
+            .map(|mut event| {
+                self.handle_event(&mut event);
+                event
+            })
+        })
+    }
+}
+
+impl Default for MuxedEvents {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FuturesStream for MuxedEvents {
+    type Item = notify::Event;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut task::Context<'_>,
+    ) -> task::Poll<Option<Self::Item>> {
+        use core::future::Future;
+        use futures_util::future::FutureExt;
+
+        let fut = self
+            .next_event()
+            .map(|val_opt| val_opt.map(|val_res| val_res.ok()).flatten());
+        pin_mut!(fut);
+        fut.poll(cx)
+    }
+}
+
+// TODO: maybe replace with crate `path-absolutize`
+fn absolutify(path: impl Into<PathBuf>, is_file: bool) -> PathBuf {
+    let path = path.into();
+
+    let (dir, maybe_filename) = if is_file {
+        (
+            path.parent().unwrap().to_path_buf(),
+            Some(path.file_name().unwrap().to_os_string()),
+        )
+    } else {
+        (path, None)
+    };
+
+    let dir = if let Ok(linked_dir) = dir.read_link() {
+        linked_dir
+    } else {
+        dir
+    };
+
+    let dir = if let Ok(abs_dir) = dir.canonicalize() {
+        abs_dir
+    } else {
+        dir
+    };
+
+    if let Some(filename) = maybe_filename {
+        dir.join(filename)
+    } else {
+        dir
+    }
```

Comment:
Alternate option using combinators:

```suggestion
    dir.read_link()
        .ok()
        .or_else(|| dir.canonicalize().ok())
        .or_else(|| maybe_filename.map(|filename| dir.join(filename)))
        .unwrap_or(dir);
```

Reply:
The idea here is that the operations aren't necessarily mutually-exclusive; I try to apply all operations to get the absolutest directory path possible, and then append the filename afterward to ensure it doesn't get swallowed in the process. Open to using combinators, but the `maybe_filename.map` would need to get reworked to apply to both the success and failure paths.

Reply:
Ahhhh, I definitely missed that subtlety. There's nothing wrong with what you have, I just wanted to mention the combinator route.